spacy/util.py

```diff
@@ -656,42 +656,47 @@ def decaying(start, stop, decay):
         curr -= decay
 
 
-def minibatch_by_words(examples, size, tuples=True, count_words=len, tolerance=0.2):
+def minibatch_by_words(examples, size, count_words=len, tolerance=0.2, discard_oversize=False):
     """Create minibatches of roughly a given number of words. If any examples
     are longer than the specified batch length, they will appear in a batch by
-    themselves."""
+    themselves, or be discarded if discard_oversize=True."""
     if isinstance(size, int):
         size_ = itertools.repeat(size)
     elif isinstance(size, List):
         size_ = iter(size)
     else:
         size_ = size
-    examples = iter(examples)
-    oversize = []
-    while True:
-        batch_size = next(size_)
-        tol_size = batch_size * 0.2
-        batch = []
-        if oversize:
-            example = oversize.pop(0)
-            n_words = count_words(example.doc)
+
+    target_size = next(size_)
+    tol_size = target_size * tolerance
+    batch = []
+    current_size = 0
+
+    for example in examples:
+        n_words = count_words(example.doc)
+        # add the example to the current batch if it still fits
+        if (current_size + n_words) < (target_size + tol_size):
             batch.append(example)
-            batch_size -= n_words
-        while batch_size >= 1:
-            try:
-                example = next(examples)
-            except StopIteration:
-                if batch:
-                    yield batch
-                return
-            n_words = count_words(example.doc)
-            if n_words < (batch_size + tol_size):
-                batch_size -= n_words
-                batch.append(example)
+            current_size += n_words
+        else:
+            # if the current example exceeds the batch size, it is returned separately
+            # but only if discard_oversize=False.
+            if current_size > target_size:
```

Comment:
Should this be `if n_words > target_size`?

Reply:
I think a case you might not have thought about is something like:

Target size: `1000`
Sizes: `[400, 400, 199, 3]`

We don't want to have a batch with just the `3`.

Reply:
yea, you're right on all fronts, will fix.

Reply:
RE the [400, 400, 199, 3]: let me add some tests.

Reply:
I added a test. So with target size `1000` and tol `0.2`, you'd want `[400, 400, 199, 3]` to stick together in one batch (of `[4]`), but `[400, 400, 199, 3, 250]` should be two batches, with `3` in the second batch, i.e. `[3,2]`, right? Currently the algorithm does work on the example you gave (because it always includes the tolerance in the check) but it's probably not behaving as expected on the latter example (moving `3` into the first batch instead of the second). Right?

Reply:
Note that the original implementation would have also failed on the last unit test for `[400, 400, 199, 3, 250]`, it would have also put the first 4 in the first batch.

Reply:
Ah yeah I forgot about the `tol`.

I think it's okay if the batching is greedy, i.e. I don't think it needs to move things around to make the batching as even as possible. It should fill up the current batch and then emit it.

But the problem is the batch isn't necessarily *full* when you first hit an example that doesn't fit in the current batch. And just because an example doesn't fit, doesn't mean that example should be in a batch by itself.

Reply:
I guess a fully general algorithm would look something like:

* Keep a queue of open batches, sorted inversely by length
* When you hit a new example, sort it into the largest batch it fits into
* If a batch is above the minimum size, emit it

Reply:
Ha, shoot. I had just implemented a more straightforward solution that at least satisfies the tests (if you agree with the tests!): it keeps the examples that are in the "tolerance margin" separate, and only adds them to the (previous) batch if no new batch was started ...
The added advantage being that we don't need the sort. I feel like we may have quite some added complexity on our hands with sorting?

Reply:
(Note that the old solution would have failed for the three lasts unit tests, it would have always taken the first 4 into the first batch)

Reply:
If you have a version that works, great! My algorithm definitely felt a bit painful.